nixos/modules/services/misc/taskserver/helper-tool.nix

```diff
@@ -0,0 +1,324 @@
+{ config, pkgs, lib, mkShellStr, taskd }:
+
+let
+  commandName = "nixos-taskserver";
+  mkShellName = lib.replaceStrings ["-"] ["_"];
+
+  genClientKey = ''
+    umask 0077
+    if tmpdir="$(${pkgs.coreutils}/bin/mktemp -d)"; then
+      trap "rm -rf '$tmpdir'" EXIT
+      ${pkgs.gnutls}/bin/certtool -p --bits 2048 --outfile "$tmpdir/key"
+
+      cat > "$tmpdir/template" <<-\ \ EOF
+      organization = $organisation
+      cn = ${config.server.fqdn}
+      tls_www_client
+      encryption_key
+      signing_key
+      EOF
+
+      ${pkgs.gnutls}/bin/certtool -c \
+        --load-privkey "$tmpdir/key" \
+        --load-ca-privkey "${config.dataDir}/keys/ca.key" \
+        --load-ca-certificate "${config.dataDir}/keys/ca.cert" \
+        --template "$tmpdir/template" \
+        --outfile "$tmpdir/cert"
+
+      mkdir -m 0700 -p "${config.dataDir}/keys/user/$organisation/$user"
+      chown root:root "${config.dataDir}/keys/user/$organisation/$user"
+      cat "$tmpdir/key" \
+        > "${config.dataDir}/keys/user/$organisation/$user/private.key"
+      cat "$tmpdir/cert" \
+        > "${config.dataDir}/keys/user/$organisation/$user/public.cert"
+
+      rm -rf "$tmpdir"
+      trap - EXIT
+    else
+      echo "Unable to create temporary directory for client" \
+           "certificate creation." >&2
+      exit 1
+    fi
+  '';
+
+  mkSubCommand = name: { args, description, script }: let
+    mkArg = pos: arg: "local ${arg}=\"\$${toString pos}\"";
+    mkDesc = line: "echo ${mkShellStr "    ${line}"} >&2";
+    usagePosArgs = lib.concatMapStringsSep " " (a: "<${a}>") args;
+  in ''
+    subcmd_${mkShellName name}() {
+      ${lib.concatImapStringsSep "\n  " mkArg args}
+      ${script}
+    }
+
+    usage_${mkShellName name}() {
+      echo "  ${commandName} ${name} ${usagePosArgs}" >&2
+      ${lib.concatMapStringsSep "\n  " mkDesc description}
+    }
+  '';
+
+  mkCStr = val: "\"${lib.escape ["\\" "\""] val}\"";
+
+  taskdUser = let
+    runUser = pkgs.writeText "runuser.c" ''
```

Comment:
Thought about this during sleep and I think the tool should indeed be written in one language, because adding C and using Nix to avoid limitations and/or boilerplate in the shell isn't very good if you want things to stay consistent and readable. So going to add this to the todo above.